### PR TITLE
make limited sparse printing O(1)

### PR DESCRIPTION
**Pull request by KristofferC**
By first looping from the start of the array and then from the end and merging the result. Also made some tweaks so that the number of elements above and below the triple dot separator are the same as well as better heuristics when the matrix can be printed without separator and how many elements will fit e.g (PR is lower figure):

image

Fixes https://github.com/JuliaLang/julia/issues/21063

~~We might want to have some more display tests for sparse arrays though.~~

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -134,30 +134,54 @@ function Base.show(io::IOContext, S::SparseMatrixCSC)
     end
 
     limit::Bool = get(io, :limit, false)
-    if limit
-        rows = displaysize(io)[1]
-        half_screen_rows = div(rows - 8, 2)
+    rows = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
+    rows <= 2 && return
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: this shouldn't just skip printing entirely

Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: That is what is currently done. The header is printed in another function.

Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: What do you suggest should be printed?

Comment (tkelman) [inline base/sparse/sparsematrix.jl]: I don't think you're right about that. The existing code prints something `if k < half_screen_rows || k > nnz(S)-half_screen_rows` or `k == half_screen_rows`. Are there any situations in which none of those conditions ever happens?

Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: In the screenshot above or?

Comment (tkelman) [inline base/sparse/sparsematrix.jl]: Ah that looks like a bug, something that should be fixed, not continued. It doesn't make any sense to let this ever be smaller than 2. Maybe print only the vdots if the limit is smaller than showing 3 lines

Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: Yes, I just came to the same conclusion about the dots.

**Comment by KristofferC** [comment]
updated

File: test/sparse/sparse.jl [hunk 2]
```diff
@@ -1772,6 +1772,26 @@ end
...
+                                      "  =  1.0\n  [2, 1]  =  2.0\n  [3, 2]  =  3.0\n  [4, 2]  =  4.0\n  [5, 3]  =  5.0")
+
+    show(ioc, MIME"text/plain"(), sparse(Int64[1,2,3,4,5,6], Int64[1,1,2,2,3,3], [1.0,2.0,3.0,4.0,5.0,6.0]))
+    @test String(take!(io)) ==  string("6×3 SparseMatrixCSC{Float64,Int64} with 6 stored entries:\n  [1, 1]",
+                                       "  =  1.0\n  [2, 1]  =  2.0\n  ⋮\n  [5, 3]  =  5.0\n  [6, 3]  =  6.0")
+
+    ioc = IOContext(io, :displaysize => (9, 80))
+    show(ioc, MIME"text/plain"(), sparse(Int64[1,2,3,4,5,6], Int64[1,1,2,2,3,3], [1.0,2.0,3.0,4.0,5.0,6.0]))
+    @test String(take!(io)) ==  string("6×3 SparseMatrixCSC{Float64,Int64} with 6 stored entries:\n  [1, 1]  =  1.0\n",
+       "  [2, 1]  =  2.0\n  [3, 2]  =  3.0\n  [4, 2]  =  4.0\n  [5, 3]  =  5.0\n  [6, 3]  =  6.0")
```
Comment (tkelman) [inline test/sparse/sparse.jl]: 4 space indent

**Comment by KristofferC** [comment]
This now requires O(nnz) storage for non limited printing. Thoughts about that? Worth adding a special case?

**Comment by KristofferC** [comment]
I changed it so that non limited printing now uses O(1) memory. This should be good to go imo.

File: test/sparse/sparse.jl (hunk 2 above)
Comment (tkelman) [inline test/sparse/sparse.jl]: the fix for this was removed?

Comment (KristofferC) [inline test/sparse/sparse.jl]: force pushed over the commit I made on the web interface

**Comment by tkelman** [comment]
This shouldn't need to allocate an array at all.

**Comment by tkelman** [comment]
I suspect this is going to be slower than the previous version, given the iterations where it didn't print would be very cheap.

**Comment by KristofferC** [comment]
Please provide concrete examples where the changes here would lead to a meaningful regression.

For matrices in my applications I have had to wait 5-10 seconds for the limited sparse printing to finish. This prints the same instantly. 
Are you worried about the case where you have a terminal with millions of lines such that the allocated array is a problem? Or are we only talking about `limited = false` where the array is not even used?

**Comment by tkelman** [comment]
if only a small number of elements are skipped, the array allocation would be more expensive than the cheap checks for the skipped elements. iterating in forward order would remove the need for a temporary.

**Comment by KristofferC** [comment]
To make sure I understand correctly, the worry is that for small sparse matrices, the time to print ~20 elements to terminal will now take longer time than previously? Do you think the extra code complexity in optimizing for this use case is worth it?

**Comment by tkelman** [comment]
one searchsorted call is all that's needed to find the column to start from, then large matrices wouldn't need to allocate either.

**Comment by KristofferC** [comment]
Until there is a concrete use case where the performance here actually matters, I am reluctant to spend too much time discussing optimizations to it.

**Comment by tkelman** [comment]
You merged this in a rush without actual approval from the only person who reviewed it. Please be a bit more patient.

**Comment by tkelman** [comment]
printing `speye(BigFloat, 50, 50)` as an example allocates 25% more here than on 0.6.

**Comment by ViralBShah** [comment]
Yeah, the inefficiency for the small case is fine. This is just printing a few lines. Making it work faster on the large matrices is what we really need, and seems like this addresses it.

**Comment by tkelman** [comment]
It's all limited cases. we don't generally allocate more than needed for no good reason.

**Comment by tkelman** [comment]
If someone merged their own PR that made an unnecessary copy of everything that gets printed for other array types that wouldn't stand.

**Comment by vtjnash** [comment]
> If someone ... made an unnecessary copy of everything that gets printed for other array types that wouldn't stand.

Doesn't it? This seems a fair description of how we implement showarray in most cases now. 

The main thing to fix is to make this a method of the show-MIME function, and print a more concise, unformated form in show. But that's independent of the improments in this PR.

**Comment by tkelman** [comment]
We don't build up an array with a copy of everything and reverse its order before printing it in other showarray methods from what I can see. It's wholly unnecessary here.

**Comment by StefanKarpinski** [comment]
This is for printing in the REPL. The performance doesn't matter at all. The normal array printing code prints elements multiple times just to figure out their alignment. No one has ever complained.

**Comment by tkelman** [comment]
"The performance doesn't matter at all" doesn't usually translate into "do it the slow way when there's an equally simple faster way."

**Pull request merged**